scripts/policy/frameworks/cluster/controller/main.zeek

```diff
@@ -230,91 +279,78 @@ event ClusterController::API::set_configuration_request(reqid: string, config: C
+	local insts_current: set[string];
+	local insts_new: set[string];
+
+	# A set of current instances not contained in the new config.
+	# Those will need to get dropped.
+	local insts_to_drop: set[string];
+
+	# The opposite: new instances not yet in our current set. Those we will need
+	# to establish contact with (or they with us).
+	local insts_to_add: set[string];
+
+	# The overlap: instances in both the current and new set. For those we verify
+	# that we're actually dealign with the same entities, and might need to re-
+	# connect if not.
+	local insts_to_keep: set[string];
+
+	# Alternative representation of insts_to_add, directly providing the instances.
+	local insts_to_peer: table[string] of ClusterController::Types::Instance;
+
+	# Helpful locals.
 	local inst_name: string;
 	local inst: ClusterController::Types::Instance;
-	local insts_to_add: ClusterController::Types::InstanceVec;
 
-	# A set of instances not contained in the config, that therefore needs
-	# to be dropped.
-	local insts_to_drop: set[string] = {};
+	for ( inst_name in g_instances )
+		add insts_current[inst_name];
+	for ( inst in config$instances )
+		add insts_new[inst$name];
 
-	for ( inst_name in ClusterController::instances )
-		add insts_to_drop[inst_name];
+	# Populate TODO lists for instances we need to drop, check, or add.
+	insts_to_drop = insts_current - insts_new;
+	insts_to_add = insts_new - insts_current;
+	insts_to_keep = insts_new & insts_current;
 
 	for ( inst in config$instances )
 		{
-		if ( inst$name in ClusterController::instances )
+		if ( inst$name in insts_to_add )
 			{
-			# Verify that it's actually the same endpoint. If not, kick out
-			# the old one. XXX 0.0.0.0 is a workaround until we've resolved
-			# how agents that peer with us obtain their identity.
-			if ( ( inst$host != 0.0.0.0 && inst$host != ClusterController::instances[inst$name]$host ) ||
-			     inst?$listen_port != ClusterController::instances[inst$name]?$listen_port ||
-			     ( inst?$listen_port && ClusterController::instances[inst$name]?$listen_port &&
-			       inst$listen_port != ClusterController::instances[inst$name]$listen_port ) )
-				{
-				drop_instance(ClusterController::instances[inst$name]);
-				}
-			else
-				{
-				# We know this instance, don't drop it below.
-				delete insts_to_drop[inst$name];
-				}
+			insts_to_peer[inst$name] = inst;
+			next;
 			}
 
-		if ( inst$name !in ClusterController::instances )
+		# Focus on the keepers: check for change in identity/location.
+		if ( inst$name !in insts_to_keep )
+			next;
+
+		# XXX 0.0.0.0 is a workaround until we've resolved
+		# how agents that peer with us obtain their identity.
+		# Broker ID?
+		if ( ( inst$host != 0.0.0.0 && inst$host != g_instances[inst$name]$host ) ||
+		     inst?$listen_port != g_instances[inst$name]?$listen_port ||
+		     ( inst?$listen_port && g_instances[inst$name]?$listen_port &&
+		       inst$listen_port != g_instances[inst$name]$listen_port ) )
```

Comment:
Worth factoring this out into a comparison function for clarity?

Reply:
Good one, will do. That one's unwieldy.

Reply:
Done, this moved to the new `is_instance_connectivity_change()` function in the same file.